File: c/aStarSearch.c
#include <stdio.h>
#include <stdlib.h>
#include <limits.h>

#define N 9

// structure to represetn a node in graph

struct Node
{
  int vertex;
  int g; // cost from source
  int h; // heuristic cost to destination
  int f;// total cost (g + h)
};

// function to compare two nodes
int compareNodes(const void* a, const void* b)
{
  struct Node* aNode = (struct Node*)a;
  struct Node* bNode = (struct Node*)b;
  return aNode->f - bNode->f;
}

// function to print path from src to dest
void printPath(int cameFrom[], int curr)
{
  if(curr == -1)
    return;
  printPath(cameFrom, cameFrom[curr]);
  printf("%d ", curr);
}

// a* search algorithm 

void aStarSearch(int graph[N][N], int startNode, int goalNode, int heuristic[])
{
  int openSet[N];
  int closedSet[N];
  int cameFrom[N];
  int gScore[N];
  int fScore[N];

  for(int i = 0; i < N; i++) {
    openSet[i] = 0;
    closedSet[i] = 0;
    cameFrom[i] = -1;
    gScore[i] = INT_MAX;
    fScore[i] = INT_MAX;
  }
  openSet[startNode] = 1;
  gScore[startNode] = 0;
  fScore[startNode] = heuristic[startNode];

  while(1) {
    int minIndex = -1;
    int minF = INT_MAX;
    for(int i = 0; i < N; i++) {
      if(openSet[i] && fScore[i] < minF) {
        minIndex = i;
        minF = fScore[i];
      }
    }
    if(minIndex == -1) {
      printf("no path found");
      return;
    }
    if(minIndex == goalNode) {
      printf("path found: ");
      printPath(cameFrom, goalNode);
      printf("\n");
      return;
    }

    openSet[minIndex] = 0;
    closedSet[minIndex] = 1;

    for(int i = 0; i < N; i++) {
      if(graph[minIndex][i] == 0)
        continue;
      int tentativeGScore = gScore[minIndex] + graph[minIndex][i];
      if (!openSet[i]) {
        openSet[i] = 1;
        cameFrom[i] = minIndex;
        gScore[i] = tentativeGscore;
        fScore[i] = gScore[i]+heuristic[i];
      }
      else if(tentativeGScore < gScore[i]) {
        cameFrom[i] = minIndex;
        gScore[i] = tentativeGScore;
        fScore[i] = gScore[i] + heuristic[i];
      }
    }
  }
}

int main() {
    int graph[N][N] = {
        {0, 0, 1, 0, 0, 0, 0, 0, 0},
        {0, 0, 0, 0, 0, 1, 0, 0, 0},
        {1, 0, 0, 1, 0, 0, 0, 1, 0},
        {0, 0, 1, 0, 0, 0, 1, 0, 0},
        {0, 0, 0, 0, 0, 1, 0, 0, 0},
        {0, 1, 0, 0, 1, 0, 0, 0, 0},
        {0, 0, 0, 1, 0, 0, 0, 0, 1},
        {0, 0, 1, 0, 0, 0, 0, 0, 1},
        {0, 0, 0, 0, 0, 0, 1, 1, 0}
    };

    int heuristic[N] = {
        3, 8, 6, 4, 5, 3, 4, 2, 1
    };
int startNode = 0; 
int goalNode = 8;
aStarSearch(graph, startNode, goalNode, heuristic);
return 0;
}

File: javascript/aStarSearch.js
function aStarSearch(graph, startNode, goalNode, heuristic)
{
  const openSet = new Set();
  const closedSet = new Set();
  const gScore = {};
  const fScore = {};

  openSet.add(startNode);
  gScore[startNode] = 0;
  fScore[startNode] = heuristic(startNode, goalNode);

  while(openSet.size > 0)
    {
      let current = null;
      for(const node of openSet)
        {
          if(current === null || fScore[node] < fScore[current])
          {
            current = node;
          }
        }
      if(current === goalNode)
      {
        return reconstructPath(cameFrom, current);
      }
      openSet.delete(current);
      closedSet.add(current);
      for(const neighbor in graph[current])
        {
          if(closedSet.has(neighbor)) continue;
          const tentativeGScore = gScore[current] + graph[current][neighbor];
          if(!openSet.has(neighbor))
          {
            openSet.add(neighbor);
          }
          else if(tenativeGScore >= gScore[neighbor])
          {
            continue;
          }
          cameFrom[neighbor] = current;
          gScore[neighbor] = tentativeGScore;
          fScore[neighbor] = gScore[neighbor] + heuristic(neighbor, goalNode);
        }
    }
  return null;
}
function reconstructPath(cameFrom, current)
{
  const totalPath = [current];
  while(current in cameFrom)
    {
      current = cameFrom[current];
      totalPath.unshift(current);
    }
  return totalPath;
}

// Example   
 usage:
const graph = {
  A: { B: 1, C: 4 },
  B: { A: 1, C: 2, D: 5 },
  C: { A: 4, B: 2, D: 1, E: 3 },
  D: { B: 5, C: 1, E: 1 },
  E: { C: 3, D: 1 }
};

const heuristic = (node, goal) => {
  // Euclidean distance heuristic (can be adjusted based on problem specifics)
  const dx = Math.abs(node.x - goal.x);
  const dy = Math.abs(node.y - goal.y);
  return Math.sqrt(dx * dx + dy * dy);
};

const start = 'A';
const goal = 'E';
const path = aStarSearch(graph, start, goal, heuristic);
console.log(path); //output ['A', 'C', 'D', 'E']

      
